### PR TITLE
Naive versioning of data in GS with predict and train flows

### PR DESCRIPTION
**Pull request by ugm2**

### CODE REVIEW

File: email_discriminator/workflows/initial_data_dump.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    gcs_handler.upload_original_data(file_path)
+
+    logger.info("Data pushed to GCP.")
+
+    logger.info(articles_df)
+    logger.info(articles_df["is_relevant"].value_counts())
+
+
+@flow(name="fetch_data")
+def fetch_data_flow(bucket_name):
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/initial_data_dump.py]: Flows and tasks will be untested for now

File: email_discriminator/workflows/initial_data_dump.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+)
+from email_discriminator.core.data_versioning.gcs_versioned_data_handler import (
+    GCSVersionedDataHandler,
+)
+
+BUCKET_NAME = os.getenv("BUCKET_NAME", "email-discriminator")
+
+
+@task(name="Fetch data")
+def fetch_data(builder, gcs_handler):
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/initial_data_dump.py]: Done

File: email_discriminator/workflows/predict.py [hunk 3]
```diff
@@ -1,29 +1,77 @@
...
+    """
+    logger = get_run_logger()
+    logger.info("Fetching unread emails.")
+    df = builder.create_predict_dataframe("from:dan@tldrnewsletter.com is:unread")
+    logger.info(f"Fetched unread emails with shape {df.shape}")
+    return df
+
+
+@task
+def upload_unread_emails(df: DataFrame, gcs_handler: GCSVersionedDataHandler) -> str:
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:41]: Flows and tasks will be untested for now

File: email_discriminator/workflows/predict.py [hunk 4]
```diff
@@ -1,29 +1,77 @@
...
     logger = get_run_logger()
-    logger.info(f"Loading data from {file_path}")
-    df = pd.read_csv(file_path)
+    logger.info("Uploading unread emails to Google Cloud Storage.")
+
+    # Calculate the hash of the data content.
+    data_hash = hashlib.sha256(df.to_string().encode()).hexdigest()[:10]
+
+    # Convert the DataFrame to a CSV string.
+    csv_string = df.to_csv(index=False)
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:52]: Flows and tasks will be untested for now

File: email_discriminator/workflows/predict.py [hunk 5]
```diff
@@ -49,20 +97,63 @@
...
+    Upload the predictions to Google Cloud Storage.
+    """
+    logger = get_run_logger()
+    logger.info("Uploading predictions to Google Cloud Storage")
+
+    # Calculate the hash of the data content.
+    data_hash = hashlib.sha256(df.to_string().encode()).hexdigest()[:10]
+
+    # Convert the DataFrame to a CSV string.
+    csv_string = df.to_csv(index=False)
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:118]: It adds readability

File: email_discriminator/workflows/predict.py [hunk 6]
```diff
@@ -49,20 +97,63 @@
     """
     logger = get_run_logger()
     logger.info("Making predictions")
-    return pipeline.predict(data)
+    predictions = pipeline.predict(data)
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py]: Done

File: email_discriminator/workflows/predict.py [hunk 7]
```diff
@@ -1,29 +1,77 @@
...
+@task
+def load_data(gcs_handler: GCSVersionedDataHandler, data_hash: str) -> DataFrame:
+    """
+    Load data from a CSV file in Google Cloud Storage.
+    """
+    logger = get_run_logger()
+    logger.info(f"Loading data from unlabelled data with data_hash {data_hash}")
+
+    # Download the CSV string from GCS.
+    csv_string = gcs_handler.download_unlabelled_data(data_hash)
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:70]: It adds readability

File: email_discriminator/workflows/predict.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:52]: It adds readability

File: email_discriminator/workflows/predict.py [hunk 8]
```diff
@@ -49,20 +97,63 @@
...
-    return pipeline.predict(data)
+    predictions = pipeline.predict(data)
+    # Add the predictions to the DataFrame as a new column.
+    data["predicted_is_relevant"] = predictions
+    logger.info(data)
+    return data
+
+
+@task
+def upload_predicted_data(df: DataFrame, gcs_handler: GCSVersionedDataHandler) -> str:
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/predict.py:107]: Flows and tasks will be untested for now

File: email_discriminator/workflows/initial_data_dump.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    logger.info(articles_df)
+    logger.info(articles_df["is_relevant"].value_counts())
+
+
+@flow(name="fetch_data")
+def fetch_data_flow(bucket_name):
+    logger = get_run_logger()
+    logger.info("Starting data fetching flow.")
+
+    fetcher = EmailFetcher()
```
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/initial_data_dump.py]: Done

File: email_discriminator/workflows/initial_data_dump.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/initial_data_dump.py]: Done

File: email_discriminator/workflows/initial_data_dump.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ugm2) [reply email_discriminator/workflows/initial_data_dump.py]: Done

**Review by ugm2 — COMMENTED** [review]
Done

File: email_discriminator/workflows/predict.py [hunk 10]
```diff
@@ -1,29 +1,77 @@
...
+DATA_PATH = os.getenv("DATA_PATH", "data/")
 MODEL_NAME = os.getenv("MODEL_NAME", "email_discriminator")
+BUCKET_NAME = os.getenv("BUCKET_NAME", "email-discriminator")
 
 mlflow.set_tracking_uri(MLFLOW_URI)
 
 
 @task
-def load_data(file_path: str) -> DataFrame:
+def fetch_unread_emails(builder: EmailDatasetBuilder) -> DataFrame:
```
Comment (ghost) [inline email_discriminator/workflows/predict.py:29]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py (hunk 3 above)
Comment (ghost) [inline email_discriminator/workflows/predict.py:41]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py [hunk 11]
```diff
@@ -1,29 +1,77 @@
...
+
+    # Upload the CSV string to GCS.
+    gcs_handler.upload_unlabelled_data(csv_string, data_hash)
+
+    logger.info(f"Uploaded unread emails to GCS with data_hash {data_hash}.")
+    return data_hash
+
+
+@task
+def load_data(gcs_handler: GCSVersionedDataHandler, data_hash: str) -> DataFrame:
```
Comment (ghost) [inline email_discriminator/workflows/predict.py:62]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py [hunk 12]
```diff
@@ -49,20 +97,62 @@ def predict(pipeline: PythonModel, data: DataFrame) -> pd.Series:
...
     logger.info("Making predictions")
-    return pipeline.predict(data)
+    # Add the predictions to the DataFrame as a new column.
+    data["predicted_is_relevant"] = pipeline.predict(data)
+    logger.info(data)
+    return data
+
+
+@task
+def upload_predicted_data(df: DataFrame, gcs_handler: GCSVersionedDataHandler) -> str:
```
Comment (ghost) [inline email_discriminator/workflows/predict.py:107]: This section seems to be untested, let's fix that together!

<details>
<summary><b>Pattern explanation 👈</b></summary>
As your program evolves, refactoring and modifications may subtly cause regressions. Adding well-designed tests for your features will help spot them easily and consistently. For more details, feel free to check <a href='https://www.datacamp.com/tutorial/pytest-tutorial-a-hands-on-guide-to-unit-testing'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
import pytest
from my_lib.submodule import MyClass

@pytest.mark.parametrize(
    "input_arg, expected_output",
    [
        [YOUR_INPUT, YOUR_OUTPUT],
    ],)
def test_myclass_custommethod(input_arg, expected_output):
    obj = MyClass()
    assert obj.custommethod(input_arg) == expected_output
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[untested]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py (hunk 4 above)
Comment (ghost) [inline email_discriminator/workflows/predict.py:52]: It doesn't look like we need this intermediate variable

<details>
<summary><b>Pattern explanation 👈</b></summary>
You shouldn't concede readability, but when it's possible, avoid unnecessary memory allocation. For more details, feel free to check <a href='https://www.reddit.com/r/programming/comments/aegshb/comment/edp4jyo/'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# def make_upper(input_str: str) -> str:
# 	upper_str = input_str.upper()
# 	return upper_str
def make_upper(input_str: str) -> str:
	return input_str.upper()
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[intermediate-variable]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py (hunk 7 above)
Comment (ghost) [inline email_discriminator/workflows/predict.py:70]: It doesn't look like we need this intermediate variable

<details>
<summary><b>Pattern explanation 👈</b></summary>
You shouldn't concede readability, but when it's possible, avoid unnecessary memory allocation. For more details, feel free to check <a href='https://www.reddit.com/r/programming/comments/aegshb/comment/edp4jyo/'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# def make_upper(input_str: str) -> str:
# 	upper_str = input_str.upper()
# 	return upper_str
def make_upper(input_str: str) -> str:
	return input_str.upper()
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[intermediate-variable]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py [hunk 13]
```diff
@@ -49,20 +97,62 @@ def predict(pipeline: PythonModel, data: DataFrame) -> pd.Series:
...
+    Upload the predictions to Google Cloud Storage.
+    """
+    logger = get_run_logger()
+    logger.info("Uploading predictions to Google Cloud Storage")
+
+    # Calculate the hash of the data content.
+    data_hash = hashlib.sha256(df.to_string().encode()).hexdigest()[:10]
+
+    # Convert the DataFrame to a CSV string.
+    csv_string = df.to_csv(index=False)
```
Comment (ghost) [inline email_discriminator/workflows/predict.py:118]: It doesn't look like we need this intermediate variable

<details>
<summary><b>Pattern explanation 👈</b></summary>
You shouldn't concede readability, but when it's possible, avoid unnecessary memory allocation. For more details, feel free to check <a href='https://www.reddit.com/r/programming/comments/aegshb/comment/edp4jyo/'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# def make_upper(input_str: str) -> str:
# 	upper_str = input_str.upper()
# 	return upper_str
def make_upper(input_str: str) -> str:
	return input_str.upper()
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[intermediate-variable]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

File: email_discriminator/workflows/predict.py [hunk 14]
```diff
@@ -49,20 +97,62 @@ def predict(pipeline: PythonModel, data: DataFrame) -> pd.Series:
...
     """
     logger = get_run_logger()
     logger.info("Starting prediction flow")
-    df = load_data(DATA_PATH)
+
+    # Create a GCSVersionedDataHandler instance.
+    gcs_handler = GCSVersionedDataHandler(BUCKET_NAME)
+
+    # Create a EmailDatasetBuilder instance.
+    builder = EmailDatasetBuilder(EmailFetcher(), TLDRContentParser())
```
Comment (ghost) [inline email_discriminator/workflows/predict.py:139]: It doesn't look like we need this intermediate variable

<details>
<summary><b>Pattern explanation 👈</b></summary>
You shouldn't concede readability, but when it's possible, avoid unnecessary memory allocation. For more details, feel free to check <a href='https://www.reddit.com/r/programming/comments/aegshb/comment/edp4jyo/'>this resource</a>.
</details>

#### Code example
Here is an example of how this is typically handled:
```python
# def make_upper(input_str: str) -> str:
# 	upper_str = input_str.upper()
# 	return upper_str
def make_upper(input_str: str) -> str:
	return input_str.upper()
```


***
<details>
<summary>Quack feedback loop 👍👎</summary>
<i>This comment is about <b>[intermediate-variable]</b>. Add the following reactions on this comment to let us know if:<br>
- :+1: that comment was on point<br>
- :eyes: that doesn't seem right<br>
- :-1: this isn't important for you right now<br>
- :confused: that explanation wasn't clear<br>
</i>
</details>

**Pull request merged**